File: migration/parsers/camp.py
from .utils import parse_timestamp, cast, TIMESTAMP_FORMAT

def parse_camp(row):
  date_format = "%Y-%m-%d"
  return dict(
    id = int(row[0]),
    name = row[1],
    year = int(row[2]),
    # createdAt = parse_timestamp(row[3], TIMESTAMP_FORMAT),
    # updatedAt = parse_timestamp(row[4], TIMESTAMP_FORMAT),
    registrationFee = cast(row[5], float),
    shirtPrice = cast(row[6], float),
    siblingDiscount = cast(row[7], float),
    registrationLateFee = cast(row[8], float),
    registrationOpenDate = parse_timestamp(row[9], date_format),
    registrationLateDate = parse_timestamp(row[10], date_format),
    registrationCloseDate = parse_timestamp(row[11], date_format),
    campStartDate = parse_timestamp(row[12], date_format),
    campEndDate = parse_timestamp(row[13], date_format),
    campsite = row[14],
    campsiteAddress = row[15],
    campSizeCap = cast(row[16], int),
    earlyRegDiscount = cast(row[17], float),
    earlyRegEndDate = parse_timestamp(row[18], date_format),
    preRegistrationFee = cast(row[19], float),
  )

File: schemas/src/camp.ts
import { DocumentReference, Timestamp } from "firebase/firestore"
import { Camper } from "./families"
import { Payment } from "./payments"

// Every type is defined with Partial since there's no strict schema in Firebase.
// Instead, every field is optional and nullable so these schemas are more of a guide.

// Since 2025, camp has split into multiple tracks with different registration caps.
// It's entirely possible that we'll have even more tracks in the future, so rather than
// adding new fields to the CampYear type, we'll use a map with these as the keys.
export enum CampTrack {
  YOUNGER = "Younger",
  OLDER = "Older",
}

export type CampRemainingSpots = Record<CampTrack, number>

// Helper type to make it easier to correlate a camp track with its cap and remaining spots.
export type CampTrackInfo = {
  track: CampTrack
  cap: number
  remainingSpots: number
}

// Every CampYear document has a subcollection of registrations associated with it.
export type CampYear = Partial<{
  campEndDate: Timestamp
  campStartDate: Timestamp
  campTotalSizeCap: number
  campSizeCaps: Record<CampTrack, number>

  campsite: string
  campsiteAddress: string

  // Registration for those who are pre-registered
  earlyRegDiscount: number /* DEPRECATED */
  earlyRegEndDate: Timestamp
  earlyRegOpenDate: Timestamp

  name: string

  preRegistrationFee: number
  registrationCloseDate: Timestamp
  registrationFee: number
  registrationLateDate: Timestamp
  registrationLateFee: number
  registrationOpenDate: Timestamp /* Registration for general public */

  shirtPrice: number
  siblingDiscount: number

  // Used to determine at what point do we start waitlisting campers.
  // Purposely separate from campSizeCap because we don't want cancellations to open up spots that could bypass
  // the waitlist.
  remainingSpots: CampRemainingSpots
}>

export enum RegistrationStatus {
  ACTIVE = "Active",
  CANCELLED = "Cancelled",
  WAITLIST = "Waitlist",
  PENDING_PAYMENT = "Pending Payment",
  PARTIAL_PAYMENT = "Partial Payment",
  PROCESSING_PAYMENT = "Processing Payment",
}

// Information that changes year to year
export type Registration = Partial<{
  camperName: string
  camper: DocumentReference<Camper>
  campTrack: CampTrack /* Added in 2025 */

  grade: number
  isPreRegistered: boolean
  status: RegistrationStatus
  shirtSize: string
  isReturning: boolean
  smallGroup: string
  cabinPreference: string

  createdAt: Timestamp
  updatedAt: Timestamp
  waitlistTime: Timestamp

  // Linked Payments
  // Added in 2024, not present in past registrations
  payments: DocumentReference<Payment>[]

  // User-filled notes
  additionalNotes: string | null
  internalNotes: string | null

  // Waitlist
  hasContacted: true

  // Checkout
  checkedOutTime: Timestamp
  isCheckedOut: boolean
  nameOfParentCheckedOut: string

  // Waiver Signature
  waiverFullName: string
  waiverSignature: string
  waiverSignDate: string

  // DEPRECATED in schemaVersion 2: demographics now live at
  // families/{familyId}/campers/{camperId}/private/demographics (full_admin only).
  // Kept for reading pre-2026 registrations; do not write.
  demographics: Record<string, string | string[]>

  // Test Data
  isTestData: boolean

  // Denormalized from the parent family for Firestore rule ownership checks.
  // Always lowercase (see normalizeEmail in utils.ts).
  familyEmails: string[]

  // See CURRENT_SCHEMA_VERSION in utils.ts
  schemaVersion: number
}>
